### PR TITLE
Do the SOCKS5 handshake reliably

### PR DESCRIPTION
**Pull request by vasild**
The `Socks5()` function which does the SOCKS5 handshake with the SOCKS5 proxy sends bytes to the socket without retrying partial writes.

`send(2)` may write only part of the provided data and return. In this case the caller is responsible for retrying the operation with the remaining data. Change `Socks5()` to do that. There is already a method `Sock::SendComplete()` which does exactly that, so use it in `Socks5()`.

A minor complication for this PR is that `Sock::SendComplete()` takes `std::string` argument whereas `Socks5()` has `std::vector<uint8_t>`. Thus the necessity for the first commit. It is possible to do also in other ways - convert the data in `Socks5()` to `std::string` or have just one `Sock::SendComplete()` that takes `void*` and change the callers to pass `str.data(), str.size()` or `vec.data(), vec.size()`.

This came up while testing https://github.com/bitcoin/bitcoin/pull/27375.

### CODE REVIEW

**Comment by pinheadmz** [comment]
concept ACK 608f8aa1c472da81f0c2b0db37be7dbcc932582a

Reviewed all code changes and built locally. I don't see anything wrong with this simple update, it makes a lot of sense and is implemented cleanly. Later this week I'll try to test it if I can, I know its covering a sort of intermittent failure but there might be a way to see it work in action, maybe in combination with #27375

File: src/util/sock.h [hunk 1]
```diff
@@ -223,15 +223,31 @@ class Sock
...
+     * Convenience method, equivalent to `SendComplete(data.data(), data.size(), timeout, interrupt)`.
+     */
     virtual void SendComplete(const std::string& data,
                               std::chrono::milliseconds timeout,
                               CThreadInterrupt& interrupt) const;
 
+    /**
+     * Convenience method, equivalent to `SendComplete(data.data(), data.size(), timeout, interrupt)`.
+     */
+    virtual void SendComplete(const std::vector<uint8_t>& data,
```
Comment (vasild) [inline src/util/sock.h:247]: _moving a [discussion](https://github.com/bitcoin/bitcoin/pull/28649#issuecomment-1765229184) from the main thread here_

> would it make sense to use `Span`?

Maybe. What would be the purpose? To unite the two `SendComplete(const std::string&` and `SendComplete(const std::vector<uint8_t>&` into one method?

**Review by jonatack — COMMENTED** [review]
ACK 608f8aa1c472da81f0c2b0db37be7dbcc932582a

**Comment by vasild** [comment]
`608f8aa1c4...af0fca530e`: use a `Span` for `Sock::SendComplete()`, looks better, thanks for the [suggestion](https://github.com/bitcoin/bitcoin/pull/28649#issuecomment-1765229184).

**Review by jonatack — COMMENTED** [review]
ACK af0fca530e4d8311bcb24a14c416e5ad7c30ff78

File: src/util/sock.h [hunk 2]
```diff
@@ -228,7 +228,14 @@ class Sock
...
+    virtual void SendComplete(Span<const unsigned char> data,
+                              std::chrono::milliseconds timeout,
+                              CThreadInterrupt& interrupt) const;
+
+    /**
+     * Convenience method, equivalent to `SendComplete(MakeUCharSpan(data), timeout, interrupt)`.
+     */
+    virtual void SendComplete(Span<const char> data,
                               std::chrono::milliseconds timeout,
                               CThreadInterrupt& interrupt) const;
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/util/sock.h:240]: I prefer to keep all implementation in one place and to keep the interface separate from the implementation.

**Review by pinheadmz — APPROVED** [review]
ACK af0fca530e4d8311bcb24a14c416e5ad7c30ff78

Built and ran tests locally. Confirmed the changes since my last ACK are using `Span` instead of `data` and `len`

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK af0fca530e4d8311bcb24a14c416e5ad7c30ff78
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmVC0dcACgkQ5+KYS2KJ
yToWvw/+JIyhSZGWsQz2aeiSo/NPEFUShQLyI6MB/dNiotx/WvbzwR5PkUWzg067
hJnFgdo3sS89WTLDoqpewspXgXsDcWyhaKBep8jifnxAWow+jw+DdjZO9UqnJZwN
g4T/URcXWVjYyIQY4C9OftUxfuRmIK8cUk6BDGSDNg2dxhdcaoez3DjcUHJuFFRa
vL/b96E9wuT3qHBXlCMvjvVE/hVOeSXk7sjGuGZMgtPsYl2TdH9VF1WSGh6rIAD6
Tf70Gf3smCj8samcs+TWvifqD4DTS4186Pu6F3P9WdVeyI+nZxirU2hizeQTRzmN
a8f4YcGDHNuHHSnX2es7ZelDWqfl0DfyrOVSiYr5Llac/R3Jwpb05lLw9tCktby6
AOW1Z8y7i3BAQfK6fLqJubGksr0JPiYs7HO0g3Funeu8mId7LZsSqXd6LRdgfmRj
/C8b+SlDFPd2VNa2r/X46hlBiIw+UbQ6CxO/IEkH6KbMrM7ybi3GsSrknclh9pZ3
ToUdhEkZJLjiO/O0gxQ+W+5aSJI6m79ReSndFxnbC0WHJgAz2osyxaDgr/Fljdyp
IHY0D4UjPMUVLAHwGRgbbGeePgYyO8CshI+78rIxTbbAoBS7QU0O+nDhWtd3Bfv4
+rFQnZQ8cB/J5fiINYj+T6DNmNso8IU2US4bVLvpVCyxc7rsIis=
=WJhV
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Comment by achow101** [comment]
ACK af0fca530e4d8311bcb24a14c416e5ad7c30ff78

**Pull request merged**